Themis/ViewModels/Assessment/CodeEditorViewModel.swift

```diff
@@ -78,29 +93,41 @@ class CodeEditorViewModel: ObservableObject {
         }
     }
     
+    @MainActor
     func addInlineHighlight(feedbackId: UUID) {
         if let file = selectedFile, let selectedSection = selectedSection {
-            if (inlineHighlights.contains { $0.key == file.path }) {
-                inlineHighlights[file.path]?.append(
-                    HighlightedRange(
-                        id: feedbackId.uuidString,
-                        range: selectedSection,
-                        color: UIColor.systemYellow,
-                        cornerRadius: 8
-                    )
-                )
-            } else {
-                inlineHighlights[file.path] = [
-                    HighlightedRange(
-                        id: feedbackId.uuidString,
-                        range: selectedSection,
-                        color: UIColor.systemYellow,
-                        cornerRadius: 8
-                    )
-                ]
+            appendHighlight(feedbackId: feedbackId, range: selectedSection, path: file.path)
+        }
+            undoManager.endUndoGrouping() /// undo group with addFeedback in AssessmentResult
+    }
+    
+    @MainActor
+    func loadInlineHighlight(assessmentResult: AssessmentResult, participationId: Int) async {
+        for i in 0..<assessmentResult.inlineFeedback.count {
+            // extract file path
+            let feedback = assessmentResult.inlineFeedback[i]
+            // the reference is extracted from the text since it is more detailed (includes columns and multilines)
+            if let text = feedback.text {
+                let components = text.components(separatedBy: .whitespaces)
+                let path = extractFilePath(textComponents: components)
+                let lines = extractLines(textComponents: components)
+                // assign the related file to the feedback (required for deleting)
+                if let referencedFile = allFiles.first(where: { $0.path == path }) {
+                    assessmentResult.assignFile(id: feedback.id, file: referencedFile)
+                    // this is required because the lines of a file are only availabe after the code is fetched
+                    await referencedFile.fetchCode(participationId: participationId)
+                    // indicates a multiline highight
+                    if components.count == 5 {
+                        constructHighlight(file: referencedFile, lines: lines, id: feedback.id)
+                    }
+                    // indicates a single line highlight
+                    if components.count == 7 {
+                        let columns = extractColumns(textComponents: components)
+                        constructHighlight(file: referencedFile, lines: lines, id: feedback.id, columns: columns)
+                    }
+                }
             }
         }
```

Comment:
Add `undoManager.removeAllActions()` here to remove all undo actions to prevent the user from undoing added highlights, like so:


https://user-images.githubusercontent.com/24537483/215266461-6e5a36f6-bc77-4cfd-b58f-2b364ca8f882.mp4

Reply:
Good catch 😄